File: src/main/java/me/syz/freelook/hooks/FreelookHook.java
package me.syz.freelook.hooks;

import me.syz.freelook.FreelookMod;
import me.syz.freelook.config.FreelookConfig;
import net.minecraft.client.Minecraft;

public class FreelookHook {
    public static final Minecraft mc = Minecraft.getMinecraft();

    public static boolean perspectiveToggled = false;
    public static boolean prevState = false;
    public static float cameraYaw = 0.0f;
    public static float cameraPitch = 0.0f;

    private static int previousPerspective = 0;

    public static void onPressed(boolean down) {
        if (FreelookMod.config.enabled) {
            if (down) {
                cameraYaw = mc.thePlayer.rotationYaw;
                cameraPitch = mc.thePlayer.rotationPitch;

                if (perspectiveToggled) {
                    resetPerspective();
                } else {
                    enterPerspective();
                }

                mc.renderGlobal.setDisplayListEntitiesDirty();
            } else if (FreelookConfig.hold) {
                resetPerspective();
            }
        } else if (perspectiveToggled) {
            resetPerspective();
        }
    }

    public static void enterPerspective() {
        perspectiveToggled = true;

        previousPerspective = mc.gameSettings.thirdPersonView;

        mc.gameSettings.thirdPersonView = FreelookConfig.perspective;
    }

    public static void resetPerspective() {
        perspectiveToggled = false;

        mc.gameSettings.thirdPersonView = previousPerspective;

        mc.renderGlobal.setDisplayListEntitiesDirty();
    }

    public static boolean overrideMouse() {
        if (mc.inGameHasFocus) {
            if (!perspectiveToggled) return true;

            mc.mouseHelper.mouseXYChange();

            if (FreelookConfig.yaw) handleYaw();
            if (FreelookConfig.pitch) handlePitch();

            mc.renderGlobal.setDisplayListEntitiesDirty();
        }
        return false;
    }

    public static void handleYaw() {
        float sensitivity = calculateSensitivity();
        float yaw = mc.mouseHelper.deltaX * sensitivity;

        if (FreelookConfig.invertYaw) yaw = -yaw;

        cameraYaw += yaw * 0.15f;
    }

    public static void handlePitch() {
        float sensitivity = calculateSensitivity();
        float pitch = mc.mouseHelper.deltaY * sensitivity;

        if (FreelookConfig.invertPitch) pitch = -pitch;

        cameraPitch += pitch * 0.15f;

        if (FreelookConfig.lockPitch) cameraPitch = Math.max(-90.0f, Math.min(cameraPitch, 90.0f));
    }

    public static float calculateSensitivity() {
        float sensitivity = mc.gameSettings.mouseSensitivity * 0.6f + 0.2f;
        return sensitivity * sensitivity * sensitivity * 8.0f;
    }
}

File: src/main/kotlin/com/github/chromaticforge/freelook/client/hook/FreelookHook.kt
package com.github.chromaticforge.freelook.client.hook

import com.github.chromaticforge.freelook.client.config.FreelookConfig
import org.polyfrost.oneconfig.api.hypixel.v1.HypixelUtils
import org.polyfrost.oneconfig.utils.v1.dsl.mc
import kotlin.math.pow

object FreelookHook {
    @JvmField
    var perspectiveToggled = false

    @JvmField
    var cameraYaw = 0f

    @JvmField
    var cameraPitch = 0f

    private var previousPerspective = 0

    @JvmStatic
    fun togglePerspective() {
        setPerspective(!perspectiveToggled)
    }

    @JvmStatic
    fun setPerspective(enabled: Boolean) {
        if (FreelookConfig.enabled) {
            if (perspectiveToggled == enabled) return

            cameraYaw = mc.thePlayer.rotationYaw
            cameraPitch = mc.thePlayer.rotationPitch

            if (enabled) {
                previousPerspective = mc.gameSettings.thirdPersonView
                mc.gameSettings.thirdPersonView = FreelookConfig.perspective
            } else {
                mc.gameSettings.thirdPersonView = previousPerspective
                mc.renderGlobal.setDisplayListEntitiesDirty()
            }

            perspectiveToggled = enabled
        } else {
            perspectiveToggled = false
        }
    }

    @JvmStatic
    fun overrideMouse(): Boolean {
        if (mc.inGameHasFocus) {
            if (!perspectiveToggled) return true

            if (HypixelUtils.isHypixel() || FreelookConfig.snaplook) {
                cameraYaw = mc.thePlayer.rotationYaw
                cameraPitch = mc.thePlayer.rotationPitch
                return true
            }

            mc.mouseHelper.mouseXYChange()

            val sensitivity = (mc.gameSettings.mouseSensitivity * 0.6f + 0.2f).pow(3) * 8f * 0.15f
            var yaw = mc.mouseHelper.deltaX * sensitivity
            var pitch = mc.mouseHelper.deltaY * sensitivity

            if (FreelookConfig.invertYaw) yaw = -yaw
            cameraYaw += yaw

            if (FreelookConfig.invertPitch) pitch = -pitch
            cameraPitch += pitch
            if (FreelookConfig.lockPitch) cameraPitch = cameraPitch.coerceIn(-90f, 90f)

            mc.renderGlobal.setDisplayListEntitiesDirty()
        }
        return false
    }
}